lighthouse-core/audits/third-party-summary.js

```diff
@@ -57,41 +76,90 @@ class ThirdPartySummary extends Audit {
+      byURL.set(request.url, urlStat);
     }
 
     const jsURLs = BootupTime.getJavaScriptURLs(networkRecords);
 
     for (const task of mainThreadTasks) {
-      const attributeableURL = BootupTime.getAttributableURLForTask(task, jsURLs);
-      const entity = thirdPartyWeb.getEntity(attributeableURL);
-      if (!entity) continue;
+      const attributableURL = BootupTime.getAttributableURLForTask(task, jsURLs);
 
-      const entityStats = entities.get(entity) || {...defaultEntityStat};
+      const urlStat = byURL.get(attributableURL) || {...defaultStat};
       const taskDuration = task.selfTime * cpuMultiplier;
       // The amount of time spent on main thread is the sum of all durations.
-      entityStats.mainThreadTime += taskDuration;
+      urlStat.mainThreadTime += taskDuration;
       // The amount of time spent *blocking* on main thread is the sum of all time longer than 50ms.
       // Note that this is not totally equivalent to the TBT definition since it fails to account for FCP,
       // but a majority of third-party work occurs after FCP and should yield largely similar numbers.
-      entityStats.blockingTime += Math.max(taskDuration - 50, 0);
-      entities.set(entity, entityStats);
+      urlStat.blockingTime += Math.max(taskDuration - 50, 0);
+      byURL.set(attributableURL, urlStat);
+    }
+
+    // Map each URL's stat to a particular third party entity.
+    const /* Map<ThirdPartyEntity, string[]> */ urls = new Map();
+    for (const [url, urlStat] of byURL.entries()) {
+      const entity = thirdPartyWeb.getEntity(url);
+      if (!entity) {
+        byURL.delete(url);
+        continue;
+      }
+      const entityStat = byEntity.get(entity) || {...defaultStat};
+      entityStat.transferSize += urlStat.transferSize;
+      entityStat.mainThreadTime += urlStat.mainThreadTime;
+      entityStat.blockingTime += urlStat.blockingTime;
+      byEntity.set(entity, entityStat);
+
+      const entityURLs = urls.get(entity) || [];
+      entityURLs.push(url);
+      urls.set(entity, entityURLs);
     }
 
-    return entities;
+    return {byURL, byEntity, urls};
+  }
+
+  /**
+   * @param {ThirdPartyEntity} entity
+   * @param {{byEntity: Map<ThirdPartyEntity, Summary>, byURL: Map<string, Summary>, urls: Map<ThirdPartyEntity, string[]>}} summaries
+   * @param {Summary} stats
+   * @return {Array<!URLSummary>}
+   */
+  static getSubItems(entity, summaries, stats) {
+    const entityURLs = summaries.urls.get(entity) || [];
+    let items = entityURLs
+      .map(url => /** @type {URLSummary} */ ({url, ...summaries.byURL.get(url)}))
+      // Sort by blocking time first, then transfer size to break ties.
+      .sort((a, b) => (b.blockingTime - a.blockingTime) || (b.transferSize - a.transferSize));
+
+    const runningSummary = {transferSize: 0, blockingTime: 0};
+    const minTransfer = Math.max(2000, stats.transferSize / 20);
+    const maxSubItems = Math.min(5, items.length);
```

Comment:
At first I was against the min transfer idea but the thought of filling the table with 5 tracking pixels for every ad network convinced me :) I agree we may not always want to show the subitem breakdown. I'd say we skip when no resources meeting the min thresholds (i.e. don't force always showing 1 resource + "Other").

extract all these constants to a file up top with explanations? I'd say we could up the threshold to 4 KiB `4096` too

Reply:
Will make the change. The pixels are definitely what prompted the change. Are you saying to omit the Other row entirely or keep it for large enough resources? I have mixed feelings either way--I don't think there is a perfect way to display all the information with static tables.

As an aside, it might be good to highlight the number of third party pixels somewhere.

Reply:
I also considered omitting the breakdown if there is exactly one large resource, but it might be worth keeping since the third party to script mapping might not be obvious to users.

Reply:
Yeah I like keeping it for the one large resource case, especially if there's blocking time to identify which specific script is the problem.